pkg/openwsn/sock/openwsn_sock_udp.c

```diff
@@ -0,0 +1,526 @@
+              (remote->netif != openwsn_get_netif())))) {
+            return -EINVAL;
+        }
+        dst_addr.type = ADDR_128B;
+        memcpy(&dst_addr.addr_128b, &remote->addr, LENGTH_ADDR128b);
+        dst_port = remote->port;
+    }
+    /* if remote is null then check that sock has a remote */
+    else if (sock->gen_sock.remote.family == AF_UNSPEC) {
+        return -ENOTCONN;
+    }
+    /* set destination port and address based on sock remote */
+    else {
+        dst_port = sock->gen_sock.remote.port;
+        dst_addr.type = ADDR_128B;
+        memcpy(&dst_addr.addr_128b, &sock->gen_sock.remote.addr,
+               LENGTH_ADDR128b);
+    }
+
+    if ((sock == NULL) || (sock->gen_sock.local.family == AF_UNSPEC)) {
+        /* if no sock or unbound get a random port */
+        if ((src_port = _get_dyn_port()) == 0) {
+            return -EADDRINUSE;
+        }
+    }
+    else {
+        /* set src port from sock */
+        src_port = sock->gen_sock.local.port;
+        /* set addr if sock is not null or bound */
+        src_addr.type = ADDR_128B;
+        memcpy(&src_addr.addr_128b, &sock->gen_sock.local.addr,
+               LENGTH_ADDR128b);
+    }
+
+    /* get pkt buffer and take owner ship */
+    if ((pkt = openqueue_getFreePacketBuffer(COMPONENT_SOCK_TO_UDP)) == NULL) {
+        return -ENOMEM;
+    }
+    pkt->owner = COMPONENT_SOCK_TO_UDP;
+    pkt->creator = COMPONENT_SOCK_TO_UDP;
+
+    /* set addresses and ports */
+    pkt->l3_destinationAdd.type = ADDR_128B;
+    pkt->l3_sourceAdd.type = ADDR_128B;
+    memcpy(&pkt->l3_sourceAdd.addr_128b, &src_addr.addr_128b, LENGTH_ADDR128b);
+    memcpy(&pkt->l3_destinationAdd.addr_128b, &dst_addr.addr_128b,
+           LENGTH_ADDR128b);
+    pkt->l4_destination_port = dst_port;
+    pkt->l4_sourcePortORicmpv6Type = src_port;
+
+    /* set payload */
+    if (packetfunctions_reserveHeader(&pkt, len)) {
+        openqueue_freePacketBuffer(pkt);
+        return -ENOMEM;
+    }
+    memcpy(pkt->payload, data, len);
+    pkt->l4_payload = pkt->payload;
+    pkt->l4_length = pkt->length;
+
+    /* push task to scheduler send */
+    scheduler_push_task(_sock_transmit_internal, TASKPRIO_UDP);
+
+    return len;
+}
+
+void sock_udp_close(sock_udp_t *sock)
+{
+    if (_udp_socket_list == NULL) {
+        return;
+    }
+    if (sock) {
+        LL_DELETE(_udp_socket_list, sock);
+        sock->next = NULL;
+    }
+}
+
+int sock_udp_get_local(sock_udp_t *sock, sock_udp_ep_t *ep)
+{
+    if (sock->gen_sock.local.family == AF_UNSPEC) {
+        return -EADDRNOTAVAIL;
+    }
+
+    memcpy(ep, &sock->gen_sock.local, sizeof(sock_udp_ep_t));
+
+    return 0;
+}
+
+int sock_udp_get_remote(sock_udp_t *sock, sock_udp_ep_t *ep)
+{
+    if (sock->gen_sock.remote.family == AF_UNSPEC) {
+        return -ENOTCONN;
+    }
+
+    memcpy(ep, &sock->gen_sock.remote, sizeof(sock_udp_ep_t));
+
+    return 0;
+}
+
+int sock_udp_recv(sock_udp_t *sock, void *data, size_t max_len,
```

Comment:
Just noticed this in my diff:
```suggestion
ssize_t sock_udp_recv(sock_udp_t *sock, void *data, size_t max_len,
```